Make it easier to pass diffs to list vbranch function

mtsgrd opened the pull request:
- cached function should not accept option
- pass reference to diffs rather than consuming them

<!-- GitButler Footer Boundary Top -->
---
This is **part 1 of 3 in a stack** made with GitButler:
- <kbd>&nbsp;3&nbsp;</kbd> #5940 
- <kbd>&nbsp;2&nbsp;</kbd> #5894 
- <kbd>&nbsp;1&nbsp;</kbd> #5890 👈 
<!-- GitButler Footer Boundary Bottom -->

Byron commented on crates/gitbutler-branch-actions/src/virtual.rs:
```diff
@@ -860,6 +848,104 @@ pub fn commit(
+    // Get the HEAD commit (current commit)
+    let head_commit = repo.head()?.peel_to_commit()?;
+    // Reset the index to match the HEAD commit
+    repo.reset(head_commit.as_object(), ResetType::Mixed, None)?;
```
At this time, GitButler doesn't differentiate between staged changes and unstaged ones, so resetting the index get get to the pre-staging state will be visually indistinguishable.

However, I think it would be more correct and more future-proof to go back to a previous state by writing the previous state held in memory, something like this.

```Rust
fn restore_index_on_error(repo, doit() -> Result) -> Result {
    let index = repo.index();
    let res = doit();
    if res.is_err() {
        index.write();
    }
    res
}
```

That way, changes to the index made by the user will be retained when restoring.

mtsgrd replied:
Ah, nice! Let me do it that way.

Byron commented on crates/gitbutler-branch-actions/src/virtual.rs:
```diff
@@ -860,6 +848,104 @@ pub fn commit(
+    });
+
+    let diff = diff_workdir_to_index(repo, path)?;
+    repo.apply(&diff, ApplyLocation::Index, Some(&mut apply_opts))?;
```
`gitoxide` still can't apply patches, but once it can this can be written so that it will write the `.git/index` only once.

Byron reviewed (commented):
It's great to see this happen, thank you!

(*I definitely learned something about `git2` again - it's very powerful when dealing with deltas in general and it's another area where `gitoxide` has to catch up.*)

krlvi commented:
Good work Mattias! Since this is being overhauled, it would actually be really nice to separate the run commit hooks in it's own API command. 

Separating it from the commit command will allow us to make both commands more robust & easy to test. And this way things like the `unstage_all` will only be invoked as a result of error from the hook execution.

 But what is even better is that it allows us to make a better UX - for instance, we can allow users to run a commit hook without actually committing. And even when the hook is executed just before committing, if the operation is slow, we can make it clear in the interface that it is the commit hook that is taking time.

krlvi commented:
From our conversation IRL - lets explore using an auxiliary index (not the default one)

mtsgrd committed:
Make it easier to pass diffs to list vbranch function

- cached function should not accept option
- pass reference to diffs rather than consuming them

krlvi commented on crates/gitbutler-branch-actions/src/actions.rs:
```diff
@@ -79,7 +79,7 @@ pub fn list_virtual_branches(project: &Project) -> Result<StackListResult> {
 pub fn list_virtual_branches_cached(
     project: &Project,
-    worktree_changes: Option<DiffByPathMap>,
+    diffs_by_path: DiffByPathMap,
```
The original name was more descriptive - the new name just repeats the type name

mtsgrd commented:
>From our conversation IRL - lets explore using an auxiliary index (not the default one)

I'll put that at the top of this stack!

krlvi commented:
Since this is now part of a stack, it is not clear what the purpose of this change here is. Seems trivial and I am sure I am missing something that is in the rest of the stack
